### ISSUE TITLE
velocity reference frame

### ISSUE THREAD

**Issue by dan9thsense** [beyretb/AnimalAI-Olympics#26]
Experimentation seems to indicate that the velocity observations are in the arena reference frame (moving forward while facing a corner yields a change in both the x (index 0) and z (index 2) velocity values.  Is that correct?  Are there any units associated with the velocity values, such as rows per second?

I would have thought that an agent in the wild would be able to observe its velocity in its own frame but not in the world frame.

Also, just to be sure, actions such as forward and backward are implemented as momentary accelerations in the agent's frame, is that correct?  I surmised that by seeing the effects of multiple forward inputs-- the velocity increases with each one and then gradually slows to zero after the inputs end.

**Comment by krisives** [comment]
What is a "row" in this context?

**Comment by dan9thsense** [comment]
The arena is 40 rows x 40 columns, so a velocity might be expressed as rows/sec.

**Comment by beyretb** [comment]
Hello,

We're using the Unity convention which means that:

- lengths are expressed in "units" which in our case we'll associate to 1 metre (there are no rows/columns, they are continuous values for all 3 axes)
- speeds are therefore expressed in m/s (framerate and machine independent)

Regarding the referential, each arena has its own referential for coordinates, speed is referential independent and expressed in m/s always. All arenas are static so there is no relative/absolute speed.

> I would have thought that an agent in the wild would be able to observe its velocity in its own frame but not in the world frame.

not sure I understand your point, we assume perception of velocity is the same for all agents.

> Also, just to be sure, actions such as forward and backward are implemented as momentary accelerations in the agent's frame, is that correct? 

correct

**Comment by dan9thsense** [comment]
If the velocity is in m/sec, then how big is the arena in meters?  In other words, at a velocity of 1 m/s, how long would it take for an agent to travel from one end of the arena to the other (coordinates (0,40) to (40,40) in the 40x40 arena).

I guess I didn't state the reference question clearly.  How about this-- if I get a velocity observation of (0.0, 0.0, 1.0) while the agent is facing toward the upper left corner of the arena, does that mean it is moving toward the upper left corner (that would mean we are using the agent's reference frame) or does it mean it is moving directly towards the top of the arena (that would mean we are using the arena's (or the world's) reference frame).

**Comment by ugurbolat** [comment]
> Also, just to be sure, actions such as forward and backward are implemented as momentary accelerations in the agent's frame, is that correct? I surmised that by seeing the effects of multiple forward inputs-- the velocity increases with each one and then gradually slows to zero after the inputs end.

More detail would be nice for this point, for instance; from Unity's ML-Agent's documentation? I would like to know what is the motion model of an agent?

**Comment by beyretb** [comment]
@dan9thsense 
> If the velocity is in m/sec, then how big is the arena in meters? In other words, at a velocity of 1 m/s, how long would it take for an agent to travel from one end of the arena to the other (coordinates (0,40) to (40,40) in the 40x40 arena).

The arena is 40 metres by 40 metres, it would therefore take 40 seconds in your example, sorry for the lack of clarity!

>I guess I didn't state the reference question clearly. How about this-- if I get a velocity observation of (0.0, 0.0, 1.0) while the agent is facing toward the upper left corner of the arena, does that mean it is moving toward the upper left corner (that would mean we are using the agent's reference frame) or does it mean it is moving directly towards the top of the arena (that would mean we are using the arena's (or the world's) reference frame).

OK I now understand your question, the speed at the moment is in the world's referential which as you mention doesn't make much sense from an animal perspective. We'll change that and make it agent's referential.

@ugurbolat 
> More detail would be nice for this point, for instance; from Unity's ML-Agent's documentation? I would like to know what is the motion model of an agent?

I will write a documentation about that and the above point which will itself point to the Unity documentation for the physics function we're using. We'll put that in the final release next week.
In a nutshell rotation is immediate by a fixed positive/negative angle and forward/backward motion is a constant force applied every time you send a non zero action (see [here](https://docs.unity3d.com/ScriptReference/Rigidbody.AddForce.html) and [here](https://docs.unity3d.com/ScriptReference/ForceMode.VelocityChange.html) for the exact Physics function used)

**Comment by dan9thsense** [comment]
Thanks, that clears it up.
